doc/manpages/ccndsmoketest.1
'\" t
.\"     Title: ccndsmoketest
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 01/10/2014
.\"    Manual: \ \&
.\"    Source: \ \& 0.8.2
.\"  Language: English
.\"
.TH "CCNDSMOKETEST" "1" "01/10/2014" "\ \& 0\&.8\&.2" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ccndsmoketest \- Simple program for smoke\-test of ccnd\&.
.SH "SYNOPSIS"
.sp
\fBccndsmoketest\fR [\-b] [\-t \fImsec\fR] [ (\-T | \-u) \fIhostname\fR ] ( send \fIfilename\fR | \fIsendfilename\fR\&.ccnb | recv | kill | status [\-x] | timeo \fImsec\fR ) \fI\&...\fR
.SH "DESCRIPTION"
.sp
The \fBccndsmoketest\fR is a simple program for sending and receiving data on sockets, for the purpose of testing ccnd\&.
.sp
Most of what this program does could be done with \fBnc\fR(1) (aka netcat)\&. It intentionally understands almost nothing about CCNx protocols\&.
.SH "OPTIONS"
.PP
\fB\-b\fR
.RS 4
Binary output on stdout instead of the default hex\&.
.RE
.PP
\fB\-t\fR \fImilliseconds\fR
.RS 4
Set the timeout for recv commands (default is 1000, or 1 second)\&.
.RE
.PP
\fB\-T\fR \fIhostname\fR
.RS 4
Contact the ccnd on the named host via TCP instead of the default unix\-domain socket\&.
.RE
.PP
\fB\-u\fR \fIhostname\fR
.RS 4
Contact the ccnd on the named host via UDP instead of the default unix\-domain socket\&.
.RE
.PP
\fB\-x\fR
.RS 4
Use in conjunction with status command, returns XML output on stdout
.RE
.SH "ARGUMENTS"
.sp
The remaining arguments are processed in order\&.
.PP
\fBsend\fR \fIfilename\fR
.RS 4
Send the contents of the named file\&. Use
\fI\-\fR
to read from stdin\&.
.RE
.PP
\fBsendfilename\fR\&.ccnb
.RS 4
If the file extension is
\fB\&.ccnb\fR, the
\fBsend\fR
verb may be omitted\&.
.RE
.PP
\fBrecv\fR
.RS 4
Recv one chunk (packet) and write the result to stdout (hex by default, but see \-b)\&. Times out with a message to stderr if nothing is received\&.
.RE
.PP
\fBkill\fR
.RS 4
Shuts down ccnd cleanly\&.
.RE
.PP
\fBstatus\fR
.RS 4
Used to implement
\fBccndstatus\fR(1)\&. Pass optional [\-x] argument to request XML output\&. See
CCND Status XML documentation
for more details\&.
.RE
.PP
\fBtimeo\fR \fImilliseconds\fR
.RS 4
Set the timeout for subsequent recv commands\&.
.RE
.SH "OUTPUT"
.sp
Hex by default, but binary output may be requested with \-b\&.
.SH "EXIT STATUS"
.PP
\fB0\fR
.RS 4
Success
.RE
.PP
\fBnonzero\fR
.RS 4
Failure (syntax or usage error, file not found)
.RE
.SH "SEE ALSO"
.sp
\fBccndstop\fR(1), \fBccndstatus\fR(1)
.sp
Michael Plass


NEWS
Changes between 0.8.1 and 0.8.2:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/21 for more details on CCNx 0.8.2.

- This release features a refactored strategy layer, making it straightforward to add new
  strategies in a modular way.  Alongside the traditional "default" strategy there are
  two new examples, a "parallel" strategy that sends interests on all available faces
  and a "loadsharing" strategy that sends each arriving interest to just one of the
  available faces.  Strategies may be selected by prefix.  This selection may be
  done using new subcommands to the ccndc utility.

	* Feature #100929: Implement Junit test suites to facilitate bulk testing.
	* Feature #100983: strategy layer refactoring
	* Feature #100984: strategy labels on prefixes
	* Feature #100990: Restore SYNC_NOISY to INFO
	* Feature #100992: Add test-security to standard java tests
	* Feature #100993: Add test-versioning back into standard java tests
	* Feature #100994: Remove unused cruft from build.xml
	* Feature #101036: java sync should use Deque operations instead of Stack as recommended in the javadocs
	* Feature #101039: UDP mode for java netmanagers is not tested in standard testing
	* Feature #101053: ccncatchunks2 could use improved hole-filling and interest lifetime management
	* Bug #101030: build.xml setup of ccnr environment is incorrect
	* Bug #101033: Handling of extremely large values of MaxSuffixComponents is incorrect
	* Bug #101037: ccn_seqwriter library code should not limit the block size to 4096 bytes
	* Bug #101042: CcnFileProxy uses deprecated methods
	* Bug #101046: VLC 2.1.0 release necessitates change in plugin Makefile
	* Bug #101048: Do not rely on accept(2) returned addr for recognizing a unix-domain socket
	* Bug #101052: Unbreak objtree build target
	* Bug #101057: ccn_sign_content / ccn_chk_signing_params do not properly incorporate a timestamp in a template via ccn_signing_params
	* Bug #101058: ccndc in daemon mode does not exit cleanly when ccnd goes away
	* Bug #101059: testTimeouts a little too strict
	* Bug #101060: ccnd cache lookups should take a controllable amount of work
	* Bug #101061: ccnd minimum interest lifetime problem
	* Bug #101062: some ccnd logging missing check for logging enabled before calling expensive log procedure
	* Support #101047: Prepare for ccnx-0.8.2 release

Changes between 0.8.0 and 0.8.1:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/20 for more details on CCNx 0.8.1.

- This release fixes a number of bugs, including a serious memory leak in ccnd.

	* Bug #101029: unbounded test log file growth
	* Bug #101032: ccnr should start at specified segment on checked start write
	* Bug #101034: No obvious documentation for how to start ccnr exists on website
	* Bug #101035: small leak in ccn_fetch should a realloc() fail
	* Bug #101038: Inconsistent encoding of zero-length name components
	* Bug #101043: ccnr r_store_lookup_backwards can fail
	* Bug #101044: valgrind reports memory leak in 0.8.0 ccnd
	* Bug #101049: ccnd can remove newly arrived content too quickly
	* Support #101045 Prepare for ccnx-0.8.1 release

Changes between 0.7.2 and 0.8.0:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/17 for more details on CCNx 0.8.0.

- The use of symmetric key HMACs in place of public/private key signatures
  is now an option in both libraries.  In situations where it can be used,
  it provides a much higher performance alternative.
- The content store in ccnd has been improved.  It now uses a flatname
  representation for indexing, in the same way that ccnr does.  Also, the
  underlying skiplist implementation has moved into the library, so that
  applications may take advantage of it in the future.  In the process,
  some old ccnd bugs relating to the handling of extreme values of
  FreshnessSeconds were fixed.
- The more readable escaping convention in URIs is now the default.
  If for whatever reason you prefer to use only percent-escaping in URIs, use

	CFLAGS='-g -DCCN_URI_DEFAULT_ESCAPE=4' ./configure

  before typing make.  Note that both forms of escaping are accepted as input.
- An example rc startup script is now included, but needs to be installed by
  hand.  It is coded for Ubuntu Linux, but should be easily adaptable to other
  systems.
- This release drops support for Java JDK 1.5.

  * Issues Resolved:
	* Bug #100118: ccnd does not handle FreshnessSeconds=0 correctly.
	* Bug #100119: ccnd should handle large values of FreshnessSeconds
	* Bug #100558: test-security does not pass
	* Bug #100897: timeout values  should be consistently long throughout java code 
	* Bug #100966: Incorrect processing of CCNS_SYNC_SCOPE=3
	* Bug #100977: Problem with verify in CCNInputStream when trying to retrieve key
	* Bug #100987: Non-portable use of local keyword in Bourne shell scripts
	* Bug #100998: VLC 2.1.x block_New definition removed
	* Bug #100999: java BLOCK_AFTER_FIRST_SEGMENT doesn't work in all cases
	* Bug #101000: sync_diff code should reset differencing data when aborting or completing a comparison
	* Bug #101004: QueuedContentHandler has a sync error
	* Bug #101005: CCNx Android Services build out of sync with API changes to CCNx C-lib
	* Bug #101007: ccn_resolve_version() should distinguish between no response and finding an unversioned object.
	* Bug #101010: ccnseqwriter command busy-loops when its ccnd goes away
	* Bug #101011: Compile warning when building new symmetric key code with openssl 0.9.8
	* Bug #101012: Improve recovery from node fetch timeout in java sync
	* Bug #101013: node fetch timeout in java sync implementation should be longer and settable

csrc/ccnd/ccndsmoketest.c
/**
 * @file ccndsmoketest.c
 * Simple program for smoke-test of ccnd.
 *
 * Copyright (C) 2008-2013 Palo Alto Research Center, Inc.
 *
 * This work is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License version 2 as published by the
 * Free Software Foundation.
 * This work is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details. You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */
#include <errno.h>
#include <fcntl.h>
#include <netdb.h>
#include <poll.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <sys/stat.h>
#include <sys/un.h>

#if defined(NEED_GETADDRINFO_COMPAT)
#include "getaddrinfo.h"
#include "dummyin6.h"
#endif

#include <ccn/ccnd.h>
#include <ccn/ccn_private.h>

#define CRLF "\r\n"
#define HTTPVERSION "1.1"

char rawbuf[1024*1024];

static void
printraw(char *p, int n)
{
    int i, l;
    while (n > 0) {
        l = (n > 40 ? 40 : n);
        for (i = 0; i < l; i++)
            printf(" %c", (' ' <= p[i] && p[i] <= '~') ? p[i] : '.');
        printf("\n");
        for (i = 0; i < l; i++)
            printf("%02X", (unsigned char)p[i]);
        printf("\n");
        p += l;
        n -= l;
    }
}

static int
open_local(struct sockaddr_un *sa, const char *verb)
{
    int sock;
    int res;
    
    sock = socket(AF_UNIX, SOCK_STREAM, 0);
    if (sock == -1) {
        perror("socket");
        exit(1);
    }
    res = connect(sock, (struct sockaddr *)sa, sizeof(*sa));
    if (res == -1 && errno == ENOENT) {
        /* Don't wait for startup just to shut it down */
        if (verb != NULL && 0 == strcmp(verb, "kill"))
            exit(1);
        /* Retry after a delay in case ccnd was just starting up. */
        sleep(1);
        res = connect(sock, (struct sockaddr *)sa, sizeof(*sa));
    }
    if (res == -1) {
        perror((char *)sa->sun_path);
        exit(1);
    }
    return(sock);
}

static int
open_socket(const char *host, const char *portstr, int sock_type)
{
    int res;
    int sock = 0;
    char canonical_remote[NI_MAXHOST] = "";
    struct addrinfo *addrinfo = NULL;
    struct addrinfo *myai = NULL;
    struct addrinfo hints = {0};
    
    if (portstr == NULL || portstr[0] == 0)
        portstr = CCN_DEFAULT_UNICAST_PORT;
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = sock_type;
    hints.ai_flags = 0;
#ifdef AI_ADDRCONFIG
    hints.ai_flags |= AI_ADDRCONFIG;
#endif
#ifdef AI_NUMERICSERV
    hints.ai_flags |= AI_NUMERICSERV;
#endif
    res = getaddrinfo(host, portstr, &hints, &addrinfo);
    if (res != 0 || addrinfo == NULL) {
        fprintf(stderr, "getaddrinfo(\"%s\", \"%s\", ...): %s\n",
                host, portstr, gai_strerror(res));
        exit(1);
    }
    
    res = getnameinfo(addrinfo->ai_addr, addrinfo->ai_addrlen,
                      canonical_remote, sizeof(canonical_remote), NULL, 0, 0);
    
    sock = socket(addrinfo->ai_family, addrinfo->ai_socktype, 0);
    if (sock == -1) {
        perror("socket");
        exit(1);
    }
    hints.ai_family = addrinfo->ai_family;
    hints.ai_flags = AI_PASSIVE;
#ifdef AI_NUMERICSERV
    hints.ai_flags |= AI_NUMERICSERV;
#endif
    res = getaddrinfo(NULL, NULL, &hints, &myai);
    if (myai != NULL) {
        res = bind(sock, (struct sockaddr *)myai->ai_addr, myai->ai_addrlen);
        if (res == -1) {
            perror("bind");
            exit(1);
        }
    }
    res = connect(sock, (struct sockaddr *)addrinfo->ai_addr, addrinfo->ai_addrlen);
    if (res == -1) {
        perror(canonical_remote);
        exit(1);
    }
    freeaddrinfo(addrinfo);
    if (myai != NULL)
        freeaddrinfo(myai);
    return (sock);
}

static void
send_ccnb_file(int sock, FILE *msgs, const char *filename, int is_dgram)
{
    ssize_t rawlen;
    ssize_t sres;
    int fd = 0;
    int truncated = 0;
    char onemore[1] = {0};
    
    if (strcmp(filename, "-") != 0) {
        fd = open(filename, O_RDONLY);
        if (fd == -1) {
            perror(filename);
            exit(-1);
        }
    }
    for (;;) {
        rawlen = read(fd, rawbuf, sizeof(rawbuf));
        if (rawlen == -1) {
            perror(filename);
            exit(-1);
        }
        if (rawlen == 0 && !is_dgram)
            break;
        if (is_dgram && rawlen == sizeof(rawbuf))
            truncated = read(fd, onemore, 1);
        if (truncated == 1)
            fprintf(msgs, "TRUNCATED ");
        fprintf(msgs, "send %s (%lu bytes)\n", filename, (unsigned long)rawlen);
        sres = send(sock, rawbuf, rawlen, 0);
        if (sres == -1) {
            perror("send");
            exit(1);
        }
        if (is_dgram)
            break;
    }
    if (fd != 0)
        close(fd);
}

static int
is_ccnb_name(const char *s)
{
    size_t len = strlen(s);
    return (len > 5 && 0 == strcasecmp(s + len - 5, ".ccnb"));
}

void
write_to_stream(FILE *outstream, const void *rawbuf, size_t rawlen)
{
    size_t wlen;
    wlen = fwrite(rawbuf, 1, rawlen, outstream);
    if (wlen != rawlen)
        fprintf(stderr, "short write (%ld of %ld)\n",
                (long)wlen, (long)rawlen);
    if (fflush(outstream) != 0)
        perror("fflush");
}

int
main(int argc, char **argv)
{
    struct sockaddr_un addr = {0};
    int opt;
    struct pollfd fds[1];
    int res;
    ssize_t rawlen;
    ssize_t wlen;
    int sock;
    char *filename = NULL;
    const char *portstr;
    int msec = 1000;
    int argp;
    FILE *msgs = stdout;
    FILE *outstream = NULL;
    int udp = 0;
    int tcp = 0;
    int recvloop = 0;
    int do_pclose = 0;
    const char *host = "localhost";
    const char *cmd = NULL;

    while ((opt = getopt(argc, argv, "bht:T:u:")) != -1) {
        switch (opt) {
            case 'b':
                outstream = stdout;
                msgs = stderr;
                break;
            case 't':
                msec = atoi(optarg);
                break;
            case 'u':
                udp = 1;
                host = optarg;
                break;
            case 'T':
                tcp = 1;
                host = optarg;
                break;
            case 'h':
            default:
                fprintf(stderr, "Usage %s %s\n", argv[0],
                        " [-b(inaryout)] "
                        " [-u udphost] "
                        " [-T tcphost] "
                        " [-t millisconds] "
                        " ( send <filename>"
                        " | <sendfilename>.ccnb"
                        " | recv"
                        " | kill"
                        " | status [-x]"
                        " | timeo <millisconds>"
                        " ) ...");
                exit(1);
        }
    }
    argp = optind;
    portstr = getenv(CCN_LOCAL_PORT_ENVNAME);
    ccn_setup_sockaddr_un(portstr, &addr);
    if (udp)
        sock = open_socket(host, portstr, SOCK_DGRAM);
    else if (tcp)
        sock = open_socket(host, portstr, SOCK_STREAM);
    else
        sock = open_local(&addr, argv[optind]);
    fds[0].fd = sock;
    fds[0].events = POLLIN;
    fds[0].revents = 0;
    for (argp = optind; argv[argp] != NULL; argp++) {
        if (0 == strcmp(argv[argp], "send")) {
            filename = argv[argp + 1];
            if (filename == NULL)
                filename = "-";
            else
                argp++;
            send_ccnb_file(sock, msgs, filename, udp);
        }
        else if (is_ccnb_name(argv[argp])) {
            filename = argv[argp];
            send_ccnb_file(sock, msgs, filename, udp);
        }
        else if (recvloop || 0 == strcmp(argv[argp], "recv")) {
            res = poll(fds, 1, msec);
            if (res == -1) {
                perror("poll");
                exit(1);
            }
            if (res == 0) {
                fprintf(msgs, "recv timed out after %d ms\n", msec);
                recvloop = 0;
                continue;
            }
            rawlen = recv(sock, rawbuf, sizeof(rawbuf), 0);
            if (rawlen == -1) {
                perror("recv");
                exit(1);
            }
            if (rawlen == 0)
                break;
            if (recvloop)
                argp--;
            else
                fprintf(msgs, "recv of %lu bytes\n", (unsigned long)rawlen);
            if (outstream != NULL)
                write_to_stream(outstream, rawbuf, rawlen);
            else
                printraw(rawbuf, rawlen);
        }
        else if (0 == strcmp(argv[argp], "kill")) {
            poll(fds, 1, 1);
            res = unlink((char *)addr.sun_path);
            if (res == 0) {
                res = open_socket(host, portstr, SOCK_STREAM);
                if (res != -1) {
                    write(res, " ", 1);
                    close(res);
                }
                poll(fds, 1, 5000);
                rawlen = recv(sock, rawbuf, sizeof(rawbuf), 0);
                if (rawlen == 0)
                    exit(0);
                if (rawlen > 0)
                    exit(2);
            }
            fprintf(stderr, "%s kill (%s) ", argv[0], (char *)addr.sun_path);
            perror("failed");
            exit(1);
        }
        else if (0 == strcmp(argv[argp], "timeo")) {
            if (argv[argp + 1] != NULL)
                msec = atoi(argv[++argp]);
        }
        else if (udp == 0 && 0 == strcmp(argv[argp], "status")) {
            msgs = stderr;
            if ((argv[argp + 1] != NULL) && (0 == strcmp(argv[argp + 1], "-x"))) {
                argp++;
                cmd = ("tail -n +6");
                outstream = popen(cmd, "w");
                wlen = send(sock, "GET /?f=xml " HTTPVERSION CRLF , 14, 0);
            } else {
                cmd = ("sed -e 's=[<]style .*/style[>]==g' -e 's=[<][^>]*[>]==g'|"
                   "tail -n +7");
                outstream = popen(cmd, "w");
                wlen = send(sock, "GET / " HTTPVERSION CRLF , 8, 0);
            }
            if (wlen < 0) perror("send");
            recvloop = 1;
            do_pclose = 1;
            argp--;
        }
        else {
            fprintf(stderr, "%s: unknown verb %s, try -h switch for usage\n",
                    argv[0], argv[argp]);
            exit(1);
        }
    }
    if (do_pclose)
        pclose(outstream);
    exit(0);
}
